# IIJA All Projects 

In [1]:
# ! pip install nltk
# ! pip install textblob

In [2]:
import pandas as pd
from siuba import *

from dla_utils import _dla_utils

from calitp import to_snakecase

import utils

/opt/conda/lib/python3.10/site-packages/geopandas/_compat.py:112: UserWarning: The Shapely GEOS version (3.10.3-CAPI-1.16.1) is incompatible with the GEOS version PyGEOS was compiled with (3.10.1-CAPI-1.16.0). Conversions between both will be slow.


In [3]:
pd.set_option("display.max_columns", 100)
pd.set_option('display.max_colwidth', None)


## Read Data

In [4]:
proj = utils.read_data_all()

In [5]:
proj.head(1)

fmis_transaction_date program_code   program_code_description  \
0            2022-03-03         Y230  STBG-URBANIZED >200K IIJA   

  project_number recipient_project_number  \
0        5002174              0314000222L   

                                                             project_title  \
0  BELL AVE. FROM BOLLENBACHER AVE. TO ASTORIA ST ROAD REHABILITATION (TC)   

   county_code congressional_district  improvement_type  \
0        67.00            Cong Dist 6              4.00   

  improvement_type_description  obligations_amount  \
0       4R - No Added Capacity        2,841,397.00   

  summary_recipient_defined_text_field_1_value  
0                                   L5002SACOG

In [6]:
len(proj)

1241

* need 1241 projects. 

## Use reworked `identify_agency` function
* now uses mapping instead of merge

In [7]:
full_df = utils.identify_agency(proj, 'summary_recipient_defined_text_field_1_value')

In [8]:
len(full_df)

1241

## Check `full_df`

In [9]:
full_df.sample()

fmis_transaction_date program_code      program_code_description  \
894            2022-09-01         YS30  HIGHWAY SAFETY IMP PROG IIJA   

    project_number recipient_project_number  \
894        P259002              0819000102S   

                                                                                                       project_title  \
894  IN SAN BERNARDINO COUNTY IN SAN BERNARDINO AT 0.1 MILE NORTH OF H STREET UNDERCROSSING CONSTRUCT MEDIAN BARRIER   

     county_code congressional_district  improvement_type  \
894        71.00           Cong Dist 31             21.00   

    improvement_type_description  obligations_amount  \
894                       Safety          561,200.00   

    summary_recipient_defined_text_field_1_value  implementing_agency_locode  \
894                                    S    SCAG                         NaN   

    implementing_agency  district            county_name rtpa_name mpo_name  
894          California      8.00  San Bernardino County       NaN      NaN

* using `project_number` as unique id

## Put into full function

In [10]:
full_df = utils.get_new_desc_title(full_df)

## Check new Title

In [11]:
full_df.sample(5)

fmis_transaction_date program_code        program_code_description  \
465            2022-06-16         Y301   TRANS ALTERNATIVES >200K IIJA   
84             2022-09-12         Y240  SURFAC TRNSP BLK GRTS-FLX IIJA   
362            2022-07-21         Y233     STBG IIJA OFF-SYSTEM BRIDGE   
295            2022-08-04         YS30    HIGHWAY SAFETY IMP PROG IIJA   
105            2022-09-09         YS30    HIGHWAY SAFETY IMP PROG IIJA   

    project_number recipient_project_number  \
465        5942301              0620000186L   
84         5027023              0422000486L   
362        5922117              0319000127L   
295        5448009              0819000099L   
105        5051005              0420000045L   

                                                                                                                                             project_title  \
465  ALONG GROVE AVENUE FROM PROSPECT AVENUE TO VALENTINE AVENUE, AND ALONG VALENTINE AVENUE FROM GROVE AVENUE TO NORTH AVENUE INSTALL AN ASPHALT CONCRETE   
84                                                                HEALDSBURG AVENUE BETWEEN POWELL AVENUE AND PASSALAQUA ROAD COMPLETE STREET IMPROVEMENTS   
362         CR 40 OVER CACHE CREEK, 0.12 MILES SOUTH OF SR 16 BRIDGE REPLACEMENT. REPLACE EXISTING 1 LANE BRIDGE WITH A NEW 1 LANE BRIDGE. BR#22C0091 (TC)   
295  HIGH-SPEED RURAL RESIDENTIAL AREAS BORDERED BY TWO MILE ROAD ON THE NORTH, BASELINE ROAD ON THE SOUTH, ENCELIA DRIVE TO THE WEST, AND ELM AVENUE ON T   
105   ON SILVERADO TRAIL, WITH THE PROJECT LIMITS BORDERING NAPA COUNTY, BEGINNING FROM APPROXIMATELY 175 FEET NORTH OF HOWELL MOUNTAIN ROAD AND ENDING AT   

     county_code congressional_district  improvement_type  \
465        19.00           Cong Dist 16             28.00   
84         97.00            Cong Dist 2             15.00   
362       113.00            Cong Dist 3             17.00   
295        71.00            Cong Dist 8             21.00   
105        41.00            Cong Dist 2             21.00   

                improvement_type_description  obligations_amount  \
465  Facilities for Pedestrians and Bicycles          411,733.00   
84                   Preliminary Engineering          600,000.00   
362                 Construction Engineering          769,250.00   
295                                   Safety        1,040,000.00   
105                                   Safety           66,500.00   

    summary_recipient_defined_text_field_1_value  implementing_agency_locode  \
465                                    L5942FCOG                    5,942.00   
84                                      L5920MTC                    5,920.00   
362                                   L5922SACOG                    5,922.00   
295                                    L5448SCAG                    5,448.00   
105                                     L5051MTC                    5,051.00   

    implementing_agency  district            county_name  \
465       Fresno County      6.00          Fresno County   
84        Sonoma County      4.00          Sonoma County   
362         Yolo County      3.00            Yolo County   
295    Twentynine Palms      8.00  San Bernardino County   
105           St Helena      4.00            Napa County   

                                  rtpa_name  \
465    Council of Fresno County Governments   
84   Metropolitan Transportation Commission   
362  Sacramento Area Council of Governments   
295   San Bernardino Associated Governments   
105  Metropolitan Transportation Commission   

                                           mpo_name  \
465             Council Of Fresno County Goverments   
84           Metropolitan Transportation Commission   
362          Sacramento Area Council Of Governments   
295  Southern California Association Of Governments   
105          Metropolitan Transportation Commission   

                                             project_title_new  
465   Facilities for Pedestria

In [12]:
#checking the length of entries here to make sure
len(full_df)

1241

In [13]:
full_df>>count(_.project_title_new)>>arrange(-_.n)

project_title_new   n
92                                         Planning Statewide  43
134   Road Restoration & Rehabilitation in Los Angeles County  33
143     Road Restoration & Rehabilitation in San Diego County  23
431                     Replace Signage in Los Angeles County  21
165                 Safety Improvements in Los Angeles County  19
..                                                        ...  ..
436                                Upgrade Bike Lanes in Galt   1
437         Upgrade Erosion Countermeasures in Ventura County   1
441                       Upgrade Guardrails in Jurupa Valley   1
453                 Upgrade Traffic Signals in San Bernardino   1
454                     Upgrade Traffic Signals in Santa Rosa   1

[456 rows x 2 columns]

In [14]:
len(full_df>>filter(_.project_title_new.str.contains("in California")))

0

In [15]:
full_df>>filter(_.project_title_new.str.contains("in California"))>>count(_.project_title_new)

Empty DataFrame
Columns: [project_title_new, n]
Index: []

In [16]:
full_df>>filter(_.project_title_new=="Road Restoration & Rehabilitation in California")

Empty DataFrame
Columns: [fmis_transaction_date, program_code, program_code_description, project_number, recipient_project_number, project_title, county_code, congressional_district, improvement_type, improvement_type_description, obligations_amount, summary_recipient_defined_text_field_1_value, implementing_agency_locode, implementing_agency, district, county_name, rtpa_name, mpo_name, project_title_new]
Index: []

## Rework 

problem: 
* current title function `update_no_matched` overrides entries Statewide. 
* to do: change the output to be in the description column, not the title column
* then the title will be formulated/corrected with the `add_new_title` for agency name

In [17]:
testdf = full_df.copy()

In [18]:
testdf.sample()

fmis_transaction_date program_code    program_code_description  \
110            2022-08-22         Y400  CONGESTION MITIGATION IIJA   

    project_number recipient_project_number  \
110        5059236              1019000088L   

                                                                                                                                            project_title  \
110  FIVE INTERSECTIONS WITHIN THE CITY OF MODESTO AS FOLLOWS: 1.) BLUE GUM AVE/CARPENTER AVE, 2.) I ST/14TH ST, 3.) I ST./17TH ST, 4.) G ST/12TH ST, AND   

     county_code congressional_district  improvement_type  \
110        99.00           Cong Dist 10             24.00   

             improvement_type_description  obligations_amount  \
110  Traffic Management/Engineering - HOV        1,024,714.00   

    summary_recipient_defined_text_field_1_value  implementing_agency_locode  \
110                                 L5059STANCOG                    5,059.00   

    implementing_agency  district        county_name  \
110             Modesto     10.00  Stanislaus County   

                             rtpa_name                           mpo_name  \
110  Stanislaus Council of Governments  Stanislaus Council Of Governments   

                          project_title_new  
110   Traffic Management Project in Modesto

In [19]:
# def update_no_matched2(df, flag_col, desc_col, program_code_desc_col): 
#     """
#     function to itreate over projects that did not match the first time
#     using an existing project's short description of project type. 
#     """
    
#     def return_project_type(df):
        
#         if (df[flag_col] == "Project") & (df[desc_col] == "Bridge Rehabilitation") | (df[desc_col] =="Bridge Rehabilitation - No Added Capacity") | (df[desc_col] =="Bridge Rehabilitation - Added Capacity"):
#             return ("Bridge Rehabilitation")
        
#         elif (df[flag_col] == "Project") & (df[desc_col] == "Facilities for Pedestrians and Bicycles"):
#             return (df[desc_col])
        
#         elif (df[flag_col] == "Project") & (df[desc_col] == "Safety"):
#             return (df[desc_col] + " Improvements")
            
#         elif (df[flag_col] == "Project") & (df[desc_col] == "Planning "):
#             return "Project Planning" 
            
#         elif (df[flag_col] == "Project") & (df[desc_col] == "Preliminary Engineering"):
#             return (df[desc_col] + " Projects ")
        
#         elif (df[flag_col] == "Project") & (df[desc_col] == "Construction Engineering"):
#             return (df[desc_col] + " Projects")
        
#         elif (df[flag_col] == "Project") & (df[desc_col] == "4R - Restoration & Rehabilitation"):
#             return ("Road Restoration & Rehabilitation")
        
#         elif (df[flag_col] == "Project") & (df[desc_col] == "4R - Maintenance  Resurfacing"):
#             return ("Maintenance Resurfacing")
        
#         elif (df[flag_col] == "Project") & (df[desc_col] == "Bridge Replacement - Added Capacity") | (df[desc_col] == "Bridge Replacement - No Added Capacity") | (df[desc_col] == "Bridge New Construction") | (df[desc_col] == "Special Bridge"):
#             return ("Bridge Replacement")
        
#         elif (df[flag_col] == "Project") & (df[desc_col] == "Mitigation of Water Pollution due to Highway Runoff"):
#             return (df[desc_col])
        
#         elif (df[flag_col] == "Project") & (df[desc_col] == "4R - Added Capacity"):
#             return ("Added Roadway Capacity")
        
#         elif (df[flag_col] == "Project") & (df[desc_col] == "4R - No Added Capacity"):
#             return ("Road Construction")
        
#         elif (df[flag_col] == "Project") & (df[desc_col] == "New  Construction Roadway"):
#             return ("New Construction Roadway")
        
#         elif (df[flag_col] == "Project") & (df[desc_col] == "Traffic Management/Engineering - HOV"):
#             return ("Traffic Management Project")
        
#         elif (df[flag_col] == "Project") & (df[desc_col] == "Right of Way"):
#             return (df[desc_col] + " Project")
        
#         elif (df[flag_col] == "Project") & (df[program_code_desc_col]== "NATIONAL HIGHWAY PERF IIJA"): 
#             return ("National Highway System Support") 
        
#         elif (df[flag_col] == "Project") & (df[desc_col] != "Other"):
#             return (df[desc_col])
    
#         else:
#             return df[flag_col] 

#         return df

#     df['project_type'] = df.apply(return_project_type, axis = 1)

#     return df

In [20]:
## move off to utils
# #deconstruct this function
# def get_new_desc_title(df):
#     '''
#     function takes the unique project ids and uses the title functions above to create a
#     unique title for each project (each project can have multiple phases).
#     this function then maps back to the original df
    
#     note: we can only use this function once in the notebook. second time running will throw an error. 
#     '''
    
#     proj_one_row = df.groupby(['project_number', 'project_title'])['obligations_amount'].max().reset_index()
#     merge_cols = ['project_number','project_title','obligations_amount']
    
#     proj_unique = (pd.merge(proj_one_row, df, how='left', on=merge_cols))
    
#     #add descriptions
#     proj_unique_cat = utils.add_description(proj_unique, 'project_title')
    
#     #remove project method column values so that the title function wont double count
#     proj_unique_cat.loc[proj_unique_cat['project_type'] == 'Project', 'project_method'] = ""
    
#      #update for the projects not in the first round of descriptions
#     proj_unique_cat_title = utils.update_no_matched(proj_unique_cat, "project_type", 'improvement_type_description', 'program_code_description')
    
#     #add title - second round to account for statewide projects
#     proj_unique_cat_title = utils.add_new_title(proj_unique_cat, "project_method", "project_type", "implementing_agency", "county_name")
    
#     # rename new title one
#     proj_unique_cat_title = proj_unique_cat_title.rename(columns={'project_name_new':'project_title_new'})
    
#     #map the title back to df
#     proj_title_mapping = (dict(proj_unique_cat_title[['project_number', 'project_title_new']].values))
    
#     df['project_title_new'] = df.project_number.map(proj_title_mapping)

#     return df


In [21]:
# Testing reworked function

In [22]:
# testdf =.get_new_desc_title(testdf)

In [23]:
# testdf

### Deconstructing Function

In [24]:
# proj_one_row = testdf.groupby(['project_number', 'project_title'])['obligations_amount'].max().reset_index()
# merge_cols = ['project_number','project_title','obligations_amount']
    
# proj_unique = (pd.merge(proj_one_row, testdf, how='left', on=merge_cols))

In [25]:
# proj_unique_cat = utils.add_description(proj_unique, 'project_title')

In [26]:
# proj_unique_cat.loc[proj_unique_cat['project_type'] == 'Project', 'project_method'] = ""

In [27]:
# proj_unique_cat = update_no_matched2(proj_unique_cat, "project_type", 'improvement_type_description', 'program_code_description')

In [28]:
# proj_unique_cat_title = utils.add_new_title(proj_unique_cat, "project_method", "project_type", "implementing_agency", "county_name")

In [29]:
# proj_unique_cat_title

In [30]:
# checking to make sure the project names have changed
# proj_unique_cat_title>>count(_.project_name_new)>>arrange(-_.n)

In [31]:
# worked! adjusting utils functions 

## Export data frame with new titles

In [32]:
#GCS_FILE_PATH  = 'gs://calitp-analytics-data/data-analyses/dla/dla-iija'

In [33]:
#test_df = utils.title_column_names(full_df)

In [34]:
#test_df.to_csv(f"{GCS_FILE_PATH}/FMIS_projects_all.csv")

In [35]:
full_df.sample(4)

fmis_transaction_date program_code      program_code_description  \
29             2022-02-10         YS30  HIGHWAY SAFETY IMP PROG IIJA   
535            2022-07-18         Y230     STBG-URBANIZED >200K IIJA   
755            2022-07-06         Y001    NATIONAL HIGHWAY PERF IIJA   
829            2022-05-06         Y001    NATIONAL HIGHWAY PERF IIJA   

    project_number recipient_project_number  \
29         5005165              0422000006L   
535        6084285              0422000500L   
755        P060161              0716000080S   
829        P110023              0716000206S   

                                                                                                                                             project_title  \
29                                                                  MONTEREY ROAD FROM COTTLE ROAD TO METCALF ROAD UPGRADE GUARDRAILS TO CURRENT STANDARDS   
535        BAY AREA REGIONAL: REGIONAL PLANNING ACTIVITIES AND PLANNING, PROGRAMMING AND MONITORING (PPM). PRIOR YEAR FUNDING PROGRAMMED ON REG170001 (TC)   
755  IN AND NEAR DIAMOND BAR, FROM FULLERTON ROAD TO DIAMOND BAR BOULEVARD. CONSTRUCT AND INSTALL STORMWATER TREATMENT BEST MANAGEMENT PRACTICES (BMPS) IN   
829  IN VARIOUS CITIES, FROM 9TH STREET/GAFFEY STREET TO END OF FREEWAY. REPLACE SIGN PANELS WITH RETROREFLECTIVE SHEETING, REMOVE OVERHEAD LIGHTING AND C   

     county_code congressional_district  improvement_type  \
29         85.00           Cong Dist 19             17.00   
535         1.00           Cong Dist 17             44.00   
755        37.00           Cong Dist 39             17.00   
829        37.00           Cong Dist 37             44.00   

    improvement_type_description  obligations_amount  \
29      Construction Engineering          150,000.00   
535                        Other       16,383,916.54   
755     Construction Engineering        1,697,780.50   
829                        Other          551,419.28   

    summary_recipient_defined_text_field_1_value  implementing_agency_locode  \
29                                      L5005MTC                    5,005.00   
535                                     L6084MTC                    6,084.00   
755                                    S    SCAG                         NaN   
829                                    S    SCAG                         NaN   

                        implementing_agency  district         county_name  \
29                                 San Jose      4.00  Santa Clara County   
535  Metropolitan Transportation Commission      4.00        Multi-County   
755                              California      7.00  Los Angeles County   
829                              California      7.00  Los Angeles County   

                                  rtpa_name  \
29   Metropolitan Transportation Commission   
535                                NON-RTPA   
755                                     NaN   
829                                     NaN   

                                   mpo_name  \
29   Metropolitan Transportation Commission   
535  Metropolitan Transportation Commission   
755                                     NaN   
829                                     NaN   

                                                           project_title_new  
29                                            Upgrade Guardrails in San Jose  
535   Regional Planning Activities in Metropolitan Transportation Commission  
755                     Install Storm Water Mitigation in Los Angeles County  
829                                    Replace Signage in Los Angeles County